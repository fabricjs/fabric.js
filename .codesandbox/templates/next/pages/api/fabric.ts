import { NextApiRequest, NextApiResponse } from 'next';
import * as fabric from 'fabric/node';

export default (req: NextApiRequest, res: NextApiResponse) => {
  try {
    const canvas = new fabric.StaticCanvas(null, { width: 100, height: 100 });
    const rect = new fabric.Rect({ width: 20, height: 50, fill: '#ff0000' });
    const text = new fabric.Text('fabric.js', { fill: 'blue', fontSize: 24 });
    canvas.add(rect, text);
    canvas.renderAll();
    res.setHeader('Content-Type', 'image/png');
    res.setHeader('Content-Disposition', 'attachment; filename="fabric.png"');
    canvas.createPNGStream().pipe(res);
  } catch (err) {
    res.status(404).json({ statusCode: 404, message: err.message });
  }
};
