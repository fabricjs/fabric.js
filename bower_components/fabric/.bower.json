{
  "name": "fabric.js",
  "version": "1.4.4",
  "homepage": "http://fabricjs.com",
  "authors": [
    "kangax",
    "Kienz"
  ],
  "description": "Object model for HTML5 canvas, and SVG-to-canvas parser. Backed by jsdom and node-canvas.",
  "main": "./dist/fabric.min.js",
  "keywords": [
    "canvas",
    "graphic",
    "graphics",
    "SVG",
    "node-canvas",
    "parser",
    "HTML5",
    "object model"
  ],
  "license": "MIT",
  "scripts": [
    "./dist/fabric.js"
  ],
  "_release": "1.4.4",
  "_resolution": {
    "type": "version",
    "tag": "v1.4.4",
    "commit": "3c656e9c95ee705e83a9f49fe6cb6862270cf9a9"
  },
  "_source": "git://github.com/kangax/fabric.js.git",
  "_target": "~1.4.4",
  "_originalSource": "fabric",
  "_direct": true
}