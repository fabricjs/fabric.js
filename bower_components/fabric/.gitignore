node_modules
.DS_Store
before_commit
