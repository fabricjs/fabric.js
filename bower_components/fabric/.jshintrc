{
  "globals": {
    "define":             true,
    "exports":            true,
    "fabric":             true,
    "Cufon":              true,
    "Event":              true,
    "G_vmlCanvasManager": true,
    "ActiveXObject":      true
  },

  "node":       true,
  "es5":        false,
  "browser":    true,

  "boss":       false,
  "curly":      false,
  "debug":      false,
  "devel":      false,
  "eqeqeq":     true,
  "eqnull":     true,
  "evil":       true,
  "expr":       true,
  "forin":      false,
  "immed":      true,
  "laxbreak":   true,
  "loopfunc":   true,
  "multistr":   true,
  "newcap":     true,
  "noarg":      true,
  "noempty":    false,
  "nonew":      false,
  "nomen":      false,
  "onevar":     false,
  "plusplus":   false,
  "regexp":     false,
  "undef":      true,
  "sub":        true,
  "strict":     false,
  "white":      false,
  "unused":     true,
  "lastsemic":  true,

  // "maxparams":  4
  // "maxcomplexity":  7
  // "maxlen":         100
  "maxdepth":       4,
  "maxstatements":  30
}
