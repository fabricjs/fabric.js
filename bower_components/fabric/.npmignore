src/
lib/
dist/all.min.js
dist/all.min.js.gz
dist/fabric.min.js
dist/fabric.min.js.gz
.DS_Store
HEADER.js
build.js
