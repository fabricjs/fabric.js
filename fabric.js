/*! Fabric.js Copyright 2010, Bitsonnet (Juriy Zaytsev, Maxim Chernyak) */

if (typeof console == 'undefined') {
  var console = {
    log: function() { },
    warn: function() { }
  };
}

var fabric = fabric || { version: 0.1 };

//= require "lib/json2"

//= require "src/util"
//= require "src/parser"
                 
//= require "src/point.class"
//= require "src/intersection.class"

//= require "src/color.class"

//= require "src/element.class"

//= require "src/object.class"
//= require "src/line.class"
//= require "src/circle.class"
//= require "src/triangle.class"
//= require "src/ellipse.class"
//= require "src/rect.class"
//= require "src/polyline.class"
//= require "src/polygon.class"

//= require "src/path.class"
//= require "src/path_group.class"

//= require "src/group.class"

//= require "src/text.class"

//= require "src/image.class"