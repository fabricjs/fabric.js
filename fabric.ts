export { getEnv, getFabricDocument, getFabricWindow, setEnv } from './src/env';
export { cache } from './src/cache';
export { VERSION as version, iMatrix } from './src/constants';
export { config } from './src/config';
export { classRegistry } from './src/ClassRegistry';
export { runningAnimations } from './src/util/animation/AnimationRegistry';

export * from './src/typedefs';

export * from './src/EventTypeDefs';
export { Observable } from './src/Observable';

export type {
  TCanvasSizeOptions,
  TSVGExportOptions,
} from './src/canvas/StaticCanvas';
export type { StaticCanvasOptions } from './src/canvas/StaticCanvasOptions';
export { StaticCanvas } from './src/canvas/StaticCanvas';
export { Canvas } from './src/canvas/Canvas';
export type { CanvasOptions } from './src/canvas/CanvasOptions';
export { CanvasDOMManager } from './src/canvas/DOMManagers/CanvasDOMManager';
export { StaticCanvasDOMManager } from './src/canvas/DOMManagers/StaticCanvasDOMManager';

export type { XY } from './src/Point';
export { Point } from './src/Point';
export type { IntersectionType } from './src/Intersection';
export { Intersection } from './src/Intersection';
export { Color } from './src/color/Color';
export * from './src/color/typedefs';

export * from './src/gradient';
export * from './src/Pattern';
export { Shadow } from './src/Shadow';
export type { SerializedShadowOptions } from './src/Shadow';

export { BaseBrush } from './src/brushes/BaseBrush';
export * from './src/brushes/typedefs';

export { PencilBrush } from './src/brushes/PencilBrush';
export { CircleBrush } from './src/brushes/CircleBrush';
export { SprayBrush } from './src/brushes/SprayBrush';
export { PatternBrush } from './src/brushes/PatternBrush';

export {
  FabricObject,

  /**
   * @deprecated Due to a naming conflict with the
   * {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object JS API},
   * `fabric.Object` has been renamed to `FabricObject`
   *
   * @example
   * import { Object } from 'fabric'; // deprecated
   * import { FabricObject } from 'fabric'; // migration path
   *
   */
  FabricObject as Object,
} from './src/shapes/Object/FabricObject';

export type {
  TFabricObjectProps,
  FabricObjectProps,
  SerializedObjectProps,
} from './src/shapes/Object/types';
export type { SerializedLineProps } from './src/shapes/Line';
export { Line } from './src/shapes/Line';
export type { CircleProps, SerializedCircleProps } from './src/shapes/Circle';
export { Circle } from './src/shapes/Circle';
export { Triangle } from './src/shapes/Triangle';
export type {
  EllipseProps,
  SerializedEllipseProps,
} from './src/shapes/Ellipse';
export { Ellipse } from './src/shapes/Ellipse';
export type { RectProps, SerializedRectProps } from './src/shapes/Rect';
export { Rect } from './src/shapes/Rect';
export type { PathProps, SerializedPathProps } from './src/shapes/Path';
export { Path } from './src/shapes/Path';
export type { SerializedPolylineProps } from './src/shapes/Polyline';
export { Polyline } from './src/shapes/Polyline';
export { Polygon } from './src/shapes/Polygon';
export type {
  GraphemeBBox,
  SerializedTextProps,
  TPathAlign,
  TPathSide,
  TextProps,
} from './src/shapes/Text/Text';
export {
  FabricText,
  /**
   * @deprecated Due to a naming conflict with the
   * {@link https://developer.mozilla.org/en-US/docs/Web/API/Text/Text Web API},
   * `fabric.Text` has been renamed to `FabricText`
   *
   * @example
   * import { Text } from 'fabric'; // deprecated
   * import { FabricText } from 'fabric'; // migration path
   *
   */
  FabricText as Text,
} from './src/shapes/Text/Text';
export type {
  ITextProps,
  SerializedITextProps,
} from './src/shapes/IText/IText';
export { IText } from './src/shapes/IText/IText';
export type {
  GraphemeData,
  SerializedTextboxProps,
  TextboxProps,
} from './src/shapes/Textbox';
export { Textbox } from './src/shapes/Textbox';
export type {
  CompleteTextStyleDeclaration,
  TextStyleDeclaration,
  TextStyle,
} from './src/shapes/Text/StyledText';
export type {
  GroupEvents,
  GroupProps,
  GroupOwnProps,
  SerializedGroupProps,
} from './src/shapes/Group';
export { Group } from './src/shapes/Group';
export * from './src/LayoutManager';
export type {
  ActiveSelectionOptions,
  MultiSelectionStacking,
} from './src/shapes/ActiveSelection';
export { ActiveSelection } from './src/shapes/ActiveSelection';
export {
  FabricImage,

  /**
   * @deprecated Due to a naming conflict with the
   * {@link https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/Image Web API},
   * `fabric.Image` has been renamed to `FabricImage`
   *
   * @example
   * import { Image } from 'fabric'; // deprecated
   * import { FabricImage } from 'fabric'; // migration path
   *
   */
  FabricImage as Image,
} from './src/shapes/Image';
export type {
  ImageSource,
  SerializedImageProps,
  ImageProps,
} from './src/shapes/Image';
export { createCollectionMixin } from './src/Collection';

export * as util from './src/util';

export { loadSVGFromString } from './src/parser/loadSVGFromString';
export { loadSVGFromURL } from './src/parser/loadSVGFromURL';
export { parseSVGDocument } from './src/parser/parseSVGDocument';

// todo convert tests to jest and stop exporting those.
export { parseAttributes } from './src/parser/parseAttributes';
export { parseStyleAttribute } from './src/parser/parseStyleAttribute';
export { parsePointsAttribute } from './src/parser/parsePointsAttribute';
export { parseTransformAttribute } from './src/parser/parseTransformAttribute';
export { getCSSRules } from './src/parser/getCSSRules';
export { parseFontDeclaration } from './src/parser/parseFontDeclaration';

export { Control } from './src/controls/Control';
export * as controlsUtils from './src/controls';

export * from './src/filters';
