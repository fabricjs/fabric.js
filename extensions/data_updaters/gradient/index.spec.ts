import { describe, vitest, expect, test, beforeAll } from 'vitest';
import type {
  SerializedGradientProps,
  GradientOptions,
  ColorStop,
} from 'fabric';
import { Gradient } from 'fabric';
import { installGradientUpdater } from './index';

const oldGradientOptions: GradientOptions<'linear'> & {
  colorStops: (ColorStop & { opacity?: number })[];
} = {
  type: 'linear',
  colorStops: [
    {
      color: 'red',
      offset: 1,
      opacity: 0.5,
    },
    {
      color: 'rgba(0,0,255,0.5)',
      offset: 0.5,
    },
    {
      color: 'rgba(0,255,0,0.5)',
      offset: 0,
      opacity: 0.3,
    },
  ],
};

const oldSerializedGradient: SerializedGradientProps<'linear'> & {
  colorStops: (ColorStop & { opacity?: number })[];
} = {
  type: 'linear',
  colorStops: [
    {
      color: 'red',
      offset: 1,
      opacity: 0.5,
    },
    {
      color: 'rgba(0,0,255,0.5)',
      offset: 0.5,
    },
    {
      color: 'rgba(0,255,0,0.5)',
      offset: 0,
      opacity: 0.3,
    },
  ],
};

const addColorStopMock = vitest.fn();

const ctxMock = {
  createLinearGradient: () => ({
    addColorStop: addColorStopMock,
  }),
} as unknown as CanvasRenderingContext2D;

describe('installGradientUpdater', () => {
  describe('Without using it opacity information is lost', () => {
    test('Init gradient from options preserve old color stops', () => {
      const gradient = new Gradient(oldGradientOptions);
      expect(gradient.colorStops).toMatchSnapshot();
    });
    test('old color stops do not render the old opacity', () => {
      const gradient = new Gradient(oldGradientOptions);
      gradient.toLive(ctxMock);
      oldGradientOptions.colorStops.forEach((colorStop) => {
        expect(addColorStopMock).toHaveBeenCalledWith(
          colorStop.offset,
          colorStop.color,
        );
      });
    });
    test('FromObject will preserve the old color stop', async () => {
      const gradient = await Gradient.fromObject(oldSerializedGradient);
      expect(gradient.colorStops).toMatchSnapshot();
    });
    test('FromObject will preserve the but wont render the old opacity', async () => {
      const gradient = await Gradient.fromObject(oldSerializedGradient);
      gradient.toLive(ctxMock);
      oldGradientOptions.colorStops.forEach((colorStop) => {
        expect(addColorStopMock).toHaveBeenCalledWith(
          colorStop.offset,
          colorStop.color,
        );
      });
    });
  });
  describe('After intalling the wrapper', () => {
    beforeAll(() => {
      installGradientUpdater();
    });
    test('Init gradient from options still preserve old color stops', () => {
      const gradient = new Gradient(oldGradientOptions);
      expect(gradient.colorStops).toMatchSnapshot();
    });
    test('old color stops do not render the old opacity', () => {
      const gradient = new Gradient(oldGradientOptions);
      gradient.toLive(ctxMock);
      oldGradientOptions.colorStops.forEach((colorStop) => {
        expect(addColorStopMock).toHaveBeenCalledWith(
          colorStop.offset,
          colorStop.color,
        );
      });
    });
    test('FromObject will merge the old opacity into color', async () => {
      const gradient = await Gradient.fromObject(oldSerializedGradient);
      expect(gradient.colorStops).toMatchSnapshot();
    });
    test('FromObject will render with the new colors', async () => {
      const gradient = await Gradient.fromObject(oldSerializedGradient);
      gradient.toLive(ctxMock);
      oldGradientOptions.colorStops.forEach((colorStop, index) => {
        expect(addColorStopMock).toHaveBeenCalledWith(
          colorStop.offset,
          gradient.colorStops[index].color,
        );
      });
    });
  });
});
