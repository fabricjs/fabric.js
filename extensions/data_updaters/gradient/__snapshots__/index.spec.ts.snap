// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`installGradientUpdater > After intalling the wrapper > FromObject will merge the old opacity into color 1`] = `
[
  {
    "color": "rgba(255,0,0,0.5)",
    "offset": 1,
  },
  {
    "color": "rgba(0,0,255,0.5)",
    "offset": 0.5,
  },
  {
    "color": "rgba(0,255,0,0.3)",
    "offset": 0,
  },
]
`;

exports[`installGradientUpdater > After intalling the wrapper > Init gradient from options still preserve old color stops 1`] = `
[
  {
    "color": "red",
    "offset": 1,
    "opacity": 0.5,
  },
  {
    "color": "rgba(0,0,255,0.5)",
    "offset": 0.5,
  },
  {
    "color": "rgba(0,255,0,0.5)",
    "offset": 0,
    "opacity": 0.3,
  },
]
`;

exports[`installGradientUpdater > Without using it opacity information is lost > FromObject will preserve the old color stop 1`] = `
[
  {
    "color": "red",
    "offset": 1,
    "opacity": 0.5,
  },
  {
    "color": "rgba(0,0,255,0.5)",
    "offset": 0.5,
  },
  {
    "color": "rgba(0,255,0,0.5)",
    "offset": 0,
    "opacity": 0.3,
  },
]
`;

exports[`installGradientUpdater > Without using it opacity information is lost > Init gradient from options preserve old color stops 1`] = `
[
  {
    "color": "red",
    "offset": 1,
    "opacity": 0.5,
  },
  {
    "color": "rgba(0,0,255,0.5)",
    "offset": 0.5,
  },
  {
    "color": "rgba(0,255,0,0.5)",
    "offset": 0,
    "opacity": 0.3,
  },
]
`;
